main/resolve/test/src/mill/main/ResolveTests.scala

```diff
@@ -1110,5 +1110,54 @@ object ResolveTests extends TestSuite {
         Right(Set(_.concrete.tests.inner.foo, _.concrete.tests.inner.innerer.bar))
       )
     }
+    test("cyclicModuleRefInitError") {
+      val check = new Checker(TestGraphs.CyclicModuleRefInitError)
+      test - check.checkSeq0(
+        Seq("__"),
+        isShortError(_, "Cyclic module reference detected")
+      )
+    }
+    test("cyclicModuleRefInitError2") {
+      val check = new Checker(TestGraphs.CyclicModuleRefInitError2)
+      test - check.checkSeq0(
+        Seq("__"),
+        isShortError(_, "Cyclic module reference detected")
+      )
+    }
+    test("cyclicModuleRefInitError3") {
+      val check = new Checker(TestGraphs.CyclicModuleRefInitError3)
+      test - check.checkSeq0(
+        Seq("__"),
+        isShortError(_, "Cyclic module reference detected")
+      )
+    }
+    test("crossedCyclicModuleRefInitError") {
+      val check = new Checker(TestGraphs.CrossedCyclicModuleRefInitError)
+      test - check.checkSeq0(
+        Seq("__"),
+        isShortError(_, "Cyclic module reference detected")
+      )
+    }
+    test("nonCyclicModules") {
+      val check = new Checker(TestGraphs.NonCyclicModules)
+      test - check(
+        "__",
+        Right(Set(_.foo))
+      )
+    }
+    test("moduleRefWithNonModuleRefChild") {
+      val check = new Checker(TestGraphs.ModuleRefWithNonModuleRefChild)
+      test - check(
+        "__",
+        Right(Set(_.foo))
+      )
+    }
+    test("moduleRefCycle") {
+      val check = new Checker(TestGraphs.ModuleRefCycle)
+      test - check(
+        "__",
+        Right(Set(_.foo))
+      )
+    }
```

Comment:
Let's also add a few test cases for non-transitive resolves, both by normal wildcard `foo.bar._` or `foo.bar._.qux`, as well as no-wildcard-at-all `foo.bar.qux`. In these cases, we should raise an error if we detect a cycle, even if there isn't a risk of stackoverflow

Reply:
I have some work on it here, it's very close: https://github.com/myyk/mill/tree/resolve-need-moduleref-2

I'm stuck on cases when it's `foo.__.bar`-shaped where `foo` or `bar` can be anything, even `_`.